bentoml/cli/bento_service.py

```diff
@@ -96,20 +88,41 @@ def run_with_conda_env(bundle_path, command):
     return
 
 
-def make_bento_name_docker_compatible(name, tag):
-    """
-    Name components may contain lowercase letters, digits and separators.
-    A separator is defined as a period, one or two underscores, or one or more dashes.
+def to_valid_docker_image_name(name):
+    # https://docs.docker.com/engine/reference/commandline/tag/#extended-description
+    return name.lower().strip("._-")
 
-    A tag name must be valid ASCII and may contain lowercase and uppercase letters,
-    digits, underscores, periods and dashes. A tag name may not start with a period
-    or a dash and may contain a maximum of 128 characters.
 
-    https://docs.docker.com/engine/reference/commandline/tag/#extended-description
-    """
-    name = name.lower().strip("._-")
-    tag = tag.lstrip(".-")[:128]
-    return name, tag
+def to_valid_docker_image_version(version):
+    # https://docs.docker.com/engine/reference/commandline/tag/#extended-description
+    return version.lstrip(".-")[:128]
+
+
+def validate_tag(ctx, param, tag):  # pylint: disable=unused-argument
+    if ":" in tag:
+        name, version = tag.split(":")[:2]
+    else:
+        name, version = tag, ""
+    if to_valid_docker_image_name(name) != name:
```

Comment:
Could we use a regex to actually validate the input?  This can break if the user provides an upper case character or a non-ASCII character. see https://docs.docker.com/engine/reference/commandline/tag/#extended-description

> Name components may contain lowercase letters, digits and separators. A separator is defined as a period, one or two underscores, or one or more dashes. A name component may not start or end with a separator.

> A tag name must be valid ASCII and may contain lowercase and uppercase letters, digits, underscores, periods, and dashes. A tag name may not start with a period or a dash and may contain a maximum of 128 characters.

If it's too tricky to get this validation to match docker's rules, we can also just try to catch the `invalid argument` error when running `docker build` command, and show a proper error message to the user. Currently it shows a confusing `Broken pip` error when using an invalid docker tag:

```bash
> bentoml containerize IrisClassifier:latest -t abc___def

[2020-07-12 18:06:02,231] INFO - Getting latest version IrisClassifier:20200701222443_EC5EAB
Found Bento: /Users/chaoyu/bentoml/repository/IrisClassifier/20200701222443_EC5EAB
Bento tag was changed to be Docker compatible.
"IrisClassifier:latest" -> "abc___def:latest"
Building Docker image: abc___def
Traceback (most recent call last):
  File "/Users/chaoyu/opt/miniconda3/envs/btml-dev/lib/python3.7/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/Users/chaoyu/opt/miniconda3/envs/btml-dev/lib/python3.7/site-packages/urllib3/connectionpool.py", line 354, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/Users/chaoyu/opt/miniconda3/envs/btml-dev/lib/python3.7/http/client.py", line 1252, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Users/chaoyu/opt/miniconda3/envs/btml-dev/lib/python3.7/http/client.py", line 1298, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Users/chaoyu/opt/miniconda3/envs/btml-dev/lib/python3.7/http/client.py", line 1247, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Users/chaoyu/opt/miniconda3/envs/btml-dev/lib/python3.7/http/client.py", line 1065, in _send_output
    self.send(chunk)
  File "/Users/chaoyu/opt/miniconda3/envs/btml-dev/lib/python3.7/http/client.py", line 987, in send
    self.sock.sendall(data)
BrokenPipeError: [Errno 32] Broken pipe
```